overlord/patch/patch7.go

```diff
@@ -0,0 +1,152 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patch
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+type connStatePatch7 struct {
+	Auto             bool                   `json:"auto,omitempty"`
+	ByGadget         bool                   `json:"by-gadget,omitempty"`
+	Interface        string                 `json:"interface,omitempty"`
+	Undesired        bool                   `json:"undesired,omitempty"`
+	StaticPlugAttrs  map[string]interface{} `json:"plug-static,omitempty"`
+	DynamicPlugAttrs map[string]interface{} `json:"plug-dynamic,omitempty"`
+	StaticSlotAttrs  map[string]interface{} `json:"slot-static,omitempty"`
+	DynamicSlotAttrs map[string]interface{} `json:"slot-dynamic,omitempty"`
+}
+
+func init() {
+	patches[7] = patch7
+}
+
+// processConns updates conns map and augments it with plug-static and slot-static attributes from current snap info.
+// NOTE: missing snap or missing plugs/slots are ignored and not reported as errors as we might have stale connections
+// and ifacemgr deals with them (i.e. discards) on startup; we want to process all good slots and plugs here.
+func processConns(conns map[string]connStatePatch7, infos map[string]*snap.Info) (bool, error) {
+	var updated bool
+	for id, conn := range conns {
+		if conn.StaticPlugAttrs != nil || conn.StaticSlotAttrs != nil {
+			continue
+		}
+
+		// undesired connections have all their attributes dropped, so don't set them
+		if conn.Undesired {
+			continue
+		}
+
+		connRef, err := interfaces.ParseConnRef(id)
+		if err != nil {
+			return false, err
+		}
+
+		var ok bool
+		var plugSnapInfo, slotSnapInfo *snap.Info
+
+		// read current snap info from disk and keep it around in infos map
+		if plugSnapInfo, ok = infos[connRef.PlugRef.Snap]; !ok {
+			plugSnapInfo, err = snap.ReadCurrentInfo(connRef.PlugRef.Snap)
+			if err == nil {
+				infos[connRef.PlugRef.Snap] = plugSnapInfo
+			}
+		}
+		if slotSnapInfo, ok = infos[connRef.SlotRef.Snap]; !ok {
+			slotSnapInfo, err = snap.ReadCurrentInfo(connRef.SlotRef.Snap)
+			if err == nil {
+				infos[connRef.SlotRef.Snap] = slotSnapInfo
+			}
+		}
+
+		if slotSnapInfo != nil {
+			if slot, ok := slotSnapInfo.Slots[connRef.SlotRef.Name]; ok && slot.Attrs != nil {
+				conn.StaticSlotAttrs = slot.Attrs
+				updated = true
+			}
+		}
+
+		if plugSnapInfo != nil {
+			if plug, ok := plugSnapInfo.Plugs[connRef.PlugRef.Name]; ok && plug.Attrs != nil {
+				conn.StaticPlugAttrs = plug.Attrs
+				updated = true
+			}
+		}
+
+		conns[id] = conn
+	}
+
+	return updated, nil
+}
+
+// patch7:
+//  - add static plug and slot attributes to connections that miss them. Attributes are read from current snap info.
+func patch7(st *state.State) error {
+	infos := make(map[string]*snap.Info)
+
+	// update all pending "discard-conns" tasks as they may keep connection data in "removed".
```

Comment:
what about old-conn on connect task? does it exist only since we started storing all attrs? should we drop it again when we stop "snap connect" to run through the handler again for pre-existing connections

Reply:
Thanks for pointing this out, I missed that. This is correct, it was added with interface hooks and yes, it should be removed. I'll address remove it in a separate PR.